File: evaluation/common/memory_pipeline/window.py
"""Candidate-owned extraction windows with optional overlapping context."""

from __future__ import annotations

from dataclasses import asdict, dataclass
import re
from typing import Mapping, Sequence

from .canonical import estimate_tokens, stable_hash
from .models import (
    ConversationEpisode,
    ExtractionWindow,
    MessageRef,
    NormalizedMessage,
)


_SENTENCE_END_RE = re.compile(r".*?(?:[。！？!?；;]|\.(?:\s|$)|\n+)", re.DOTALL)


@dataclass(frozen=True)
class WindowConfig:
    max_candidate_tokens: int = 320
    max_window_tokens: int = 640
    context_before_messages: int = 2
    context_after_messages: int = 0
    tokenizer_name: str = "heuristic"
    tokenizer_version: str = "heuristic_v1"
    version: str = "window_v1"

    def __post_init__(self) -> None:
        if self.max_candidate_tokens <= 0:
            raise ValueError("max_candidate_tokens must be positive")
        if self.max_window_tokens < self.max_candidate_tokens:
            raise ValueError("max_window_tokens must be >= max_candidate_tokens")
        if self.context_before_messages < 0 or self.context_after_messages < 0:
            raise ValueError("context message counts must be non-negative")


def build_windows(
    episodes: Sequence[ConversationEpisode],
    messages_by_id: Mapping[str, NormalizedMessage],
    config: WindowConfig,
) -> list[ExtractionWindow]:
    windows: list[ExtractionWindow] = []
    for episode in episodes:
        refs: list[MessageRef] = []
        for message_id in episode.message_ids:
            try:
                message = messages_by_id[message_id]
            except KeyError as error:
                raise ValueError(f"episode references unknown message: {message_id}") from error
            refs.extend(_slice_message(message, config.max_candidate_tokens))

        groups = _pack_candidate_refs(refs, config.max_candidate_tokens)
        ref_positions = {id(ref): index for index, ref in enumerate(refs)}
        for group in groups:
            start = ref_positions[id(group[0])]
            end = ref_positions[id(group[-1])] + 1
            before = _select_context_before(
                refs[:start],
                config.context_before_messages,
            )
            after = _select_context_after(
                refs[end:],
                config.context_after_messages,
            )
            before, after = _trim_context_to_budget(
                before,
                group,
                after,
                config.max_window_tokens,
            )
            windows.append(_make_window(episode, group, before, after, config))
    return windows


def render_window(
    window: ExtractionWindow,
    messages_by_id: Mapping[str, NormalizedMessage],
) -> str:
    return "\n".join(
        [
            "<context>",
            *(_render_ref(ref, messages_by_id) for ref in window.context_before_refs),
            *(_render_ref(ref, messages_by_id) for ref in window.context_after_refs),
            "</context>",
            "",
            "<candidate>",
            *(_render_ref(ref, messages_by_id) for ref in window.candidate_refs),
            "</candidate>",
        ]
    )


def _slice_message(message: NormalizedMessage, max_tokens: int) -> list[MessageRef]:
    if estimate_tokens(message.text) <= max_tokens:
        return [MessageRef(message.id, 0, len(message.text), message.text)]

    sentence_spans = _sentence_spans(message.text)
    refs: list[MessageRef] = []
    for start, end in sentence_spans:
        refs.extend(_split_span(message, start, end, max_tokens))
    return refs


def _sentence_spans(text: str) -> list[tuple[int, int]]:
    spans: list[tuple[int, int]] = []
    cursor = 0
    for match in _SENTENCE_END_RE.finditer(text):
        if match.start() > cursor:
            spans.append((cursor, match.start()))
        spans.append((match.start(), match.end()))
        cursor = match.end()
    if cursor < len(text):
        spans.append((cursor, len(text)))
    return [(start, end) for start, end in spans if end > start]


def _split_span(
    message: NormalizedMessage,
    start: int,
    end: int,
    max_tokens: int,
) -> list[MessageRef]:
    text = message.text
    if estimate_tokens(text[start:end]) <= max_tokens:
        return [MessageRef(message.id, start, end, text[start:end])]

    refs: list[MessageRef] = []
    piece_start = start
    cursor = start
    while cursor < end:
        next_cursor = cursor + 1
        if (
            cursor > piece_start
            and estimate_tokens(text[piece_start:next_cursor]) > max_tokens
        ):
            refs.append(
                MessageRef(
                    message.id,
                    piece_start,
                    cursor,
                    text[piece_start:cursor],
                )
            )
            piece_start = cursor
        cursor = next_cursor
    if piece_start < end:
        refs.append(MessageRef(message.id, piece_start, end, text[piece_start:end]))
    return refs


def _pack_candidate_refs(
    refs: Sequence[MessageRef],
    max_tokens: int,
) -> list[tuple[MessageRef, ...]]:
    groups: list[tuple[MessageRef, ...]] = []
    current: list[MessageRef] = []
    current_tokens = 0
    for ref in refs:
        ref_tokens = estimate_tokens(ref.text)
        if current and current_tokens + ref_tokens > max_tokens:
            groups.append(tuple(current))
            current = []
            current_tokens = 0
        current.append(ref)
        current_tokens += ref_tokens
    if current:
        groups.append(tuple(current))
    return groups


def _select_context_before(
    refs: Sequence[MessageRef],
    message_limit: int,
) -> tuple[MessageRef, ...]:
    if message_limit == 0:
        return ()
    selected: list[MessageRef] = []
    seen_messages: list[str] = []
    for ref in reversed(refs):
        if ref.message_id not in seen_messages:
            if len(seen_messages) >= message_limit:
                break
            seen_messages.append(ref.message_id)
        selected.append(ref)
    return tuple(reversed(selected))


def _select_context_after(
    refs: Sequence[MessageRef],
    message_limit: int,
) -> tuple[MessageRef, ...]:
    if message_limit == 0:
        return ()
    selected: list[MessageRef] = []
    seen_messages: list[str] = []
    for ref in refs:
        if ref.message_id not in seen_messages:
            if len(seen_messages) >= message_limit:
                break
            seen_messages.append(ref.message_id)
        selected.append(ref)
    return tuple(selected)


def _trim_context_to_budget(
    before: tuple[MessageRef, ...],
    candidate: tuple[MessageRef, ...],
    after: tuple[MessageRef, ...],
    max_tokens: int,
) -> tuple[tuple[MessageRef, ...], tuple[MessageRef, ...]]:
    before_list = list(before)
    after_list = list(after)
    while _refs_tokens((*before_list, *candidate, *after_list)) > max_tokens:
        if before_list:
            before_list.pop(0)
        elif after_list:
            after_list.pop()
        else:
            break
    return tuple(before_list), tuple(after_list)


def _make_window(
    episode: ConversationEpisode,
    candidate: tuple[MessageRef, ...],
    before: tuple[MessageRef, ...],
    after: tuple[MessageRef, ...],
    config: WindowConfig,
) -> ExtractionWindow:
    config_value = asdict(config)
    window_id = "window-" + stable_hash(
        episode.scope_id,
        episode.session_id,
        [ref.to_dict() for ref in candidate],
        [ref.to_dict() for ref in before],
        [ref.to_dict() for ref in after],
        config_value,
    )
    return ExtractionWindow(
        id=window_id,
        scope_id=episode.scope_id,
        session_id=episode.session_id,
        episode_id=episode.id,
        candidate_refs=candidate,
        context_before_refs=before,
        context_after_refs=after,
        candidate_token_count=_refs_tokens(candidate),
        total_token_count=_refs_tokens((*before, *candidate, *after)),
        window_version=config.version,
    )


def _refs_tokens(refs: Sequence[MessageRef]) -> int:
    return sum(estimate_tokens(ref.text) for ref in refs)


def _render_ref(
    ref: MessageRef,
    messages_by_id: Mapping[str, NormalizedMessage],
) -> str:
    message = messages_by_id[ref.message_id]
    header = (
        f"[message_id={message.id} role={message.role} "
        f"speaker={message.speaker or '-'} time={message.timestamp or '-'} "
        f"span={ref.start_char}:{ref.end_char}]"
    )
    return f"{header}\n{ref.text}"

File: ram-a-mem/crates/memory-pipeline/src/window.rs
use std::collections::{HashMap, HashSet};

use serde::{Deserialize, Serialize};
use serde_json::{json, to_value};

use crate::canonical::{estimate_tokens, stable_hash};
use crate::error::{PipelineError, Result};
use crate::models::{ConversationEpisode, ExtractionWindow, MessageRef, NormalizedMessage};

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct WindowConfig {
    pub max_candidate_tokens: usize,
    pub max_window_tokens: usize,
    pub context_before_messages: usize,
    pub context_after_messages: usize,
    pub tokenizer_name: String,
    pub tokenizer_version: String,
    pub version: String,
}

impl Default for WindowConfig {
    fn default() -> Self {
        Self {
            max_candidate_tokens: 320,
            max_window_tokens: 640,
            context_before_messages: 2,
            context_after_messages: 0,
            tokenizer_name: "heuristic".into(),
            tokenizer_version: "heuristic_v1".into(),
            version: "window_v1".into(),
        }
    }
}

pub fn build_windows(
    episodes: &[ConversationEpisode],
    messages_by_id: &HashMap<String, NormalizedMessage>,
    config: &WindowConfig,
) -> Result<Vec<ExtractionWindow>> {
    validate_config(config)?;
    let mut windows = Vec::new();
    for episode in episodes {
        let mut all_refs = Vec::new();
        let mut eligible_ref_indices = Vec::new();
        for id in &episode.message_ids {
            let message = messages_by_id.get(id).ok_or_else(|| {
                PipelineError::InvalidInput(format!("episode references unknown message: {id}"))
            })?;
            for reference in slice_message(message, config.max_candidate_tokens) {
                if message.candidate_eligible {
                    eligible_ref_indices.push(all_refs.len());
                }
                all_refs.push(reference);
            }
        }
        let eligible_refs = eligible_ref_indices
            .iter()
            .map(|index| all_refs[*index].clone())
            .collect::<Vec<_>>();
        for (start, end) in pack_ranges(&eligible_refs, config.max_candidate_tokens) {
            let candidate = eligible_refs[start..end].to_vec();
            let candidate_indices = &eligible_ref_indices[start..end];
            let all_start = eligible_ref_indices[start];
            let all_end = eligible_ref_indices[end - 1] + 1;
            let mut before = select_before(&all_refs[..all_start], config.context_before_messages);
            before.extend(
                (all_start..all_end)
                    .filter(|index| candidate_indices.binary_search(index).is_err())
                    .map(|index| all_refs[index].clone()),
            );
            let after = select_after(&all_refs[all_end..], config.context_after_messages);
            let (before, after) = trim_context(before, &candidate, after, config.max_window_tokens);
            windows.push(make_window(episode, candidate, before, after, config));
        }
    }
    Ok(windows)
}

pub fn render_window(
    window: &ExtractionWindow,
    messages_by_id: &HashMap<String, NormalizedMessage>,
) -> Result<String> {
    let render = |reference: &MessageRef| -> Result<String> {
        let message = messages_by_id.get(&reference.message_id).ok_or_else(|| {
            PipelineError::InvalidInput(format!(
                "window references unknown message: {}",
                reference.message_id
            ))
        })?;
        Ok(format!(
            "[message_id={} role={} speaker={} time={} span={}:{}]\n{}",
            message.id,
            message.role,
            if message.speaker.is_empty() {
                "-"
            } else {
                &message.speaker
            },
            if message.timestamp.is_empty() {
                "-"
            } else {
                &message.timestamp
            },
            reference.start_char,
            reference.end_char,
            reference.text
        ))
    };
    let mut lines = vec!["<context>".to_owned()];
    for reference in window
        .context_before_refs
        .iter()
        .chain(window.context_after_refs.iter())
    {
        lines.push(render(reference)?);
    }
    lines.extend(["</context>".into(), String::new(), "<candidate>".into()]);
    for reference in &window.candidate_refs {
        lines.push(render(reference)?);
    }
    lines.push("</candidate>".into());
    Ok(lines.join("\n"))
}

fn validate_config(config: &WindowConfig) -> Result<()> {
    if config.max_candidate_tokens == 0 {
        return Err(PipelineError::InvalidInput(
            "max_candidate_tokens must be positive".into(),
        ));
    }
    if config.max_window_tokens < config.max_candidate_tokens {
        return Err(PipelineError::InvalidInput(
            "max_window_tokens must be >= max_candidate_tokens".into(),
        ));
    }
    Ok(())
}

fn slice_message(message: &NormalizedMessage, max_tokens: usize) -> Vec<MessageRef> {
    if estimate_tokens(&message.text) <= max_tokens {
        return vec![message_ref(message, 0, message.text.chars().count())];
    }
    sentence_spans(&message.text)
        .into_iter()
        .flat_map(|(start, end)| split_span(message, start, end, max_tokens))
        .collect()
}

fn sentence_spans(text: &str) -> Vec<(usize, usize)> {
    let chars = text.chars().collect::<Vec<_>>();
    let mut spans = Vec::new();
    let mut start = 0;
    let mut index = 0;
    while index < chars.len() {
        let current = chars[index];
        let mut end = None;
        if matches!(current, '。' | '！' | '？' | '!' | '?' | '；' | ';') {
            end = Some(index + 1);
        } else if current == '.' && (index + 1 == chars.len() || chars[index + 1].is_whitespace()) {
            end = Some(if index + 1 < chars.len() {
                index + 2
            } else {
                index + 1
            });
        } else if current == '\n' {
            let mut cursor = index + 1;
            while cursor < chars.len() && chars[cursor] == '\n' {
                cursor += 1;
            }
            end = Some(cursor);
        }
        if let Some(end) = end {
            if end > start {
                spans.push((start, end));
            }
            start = end;
            index = end;
        } else {
            index += 1;
        }
    }
    if start < chars.len() {
        spans.push((start, chars.len()));
    }
    spans
}

fn split_span(
    message: &NormalizedMessage,
    start: usize,
    end: usize,
    max_tokens: usize,
) -> Vec<MessageRef> {
    if estimate_tokens(&char_slice(&message.text, start, end)) <= max_tokens {
        return vec![message_ref(message, start, end)];
    }
    let mut output = Vec::new();
    let mut piece_start = start;
    let mut cursor = start;
    while cursor < end {
        let next = cursor + 1;
        if cursor > piece_start
            && estimate_tokens(&char_slice(&message.text, piece_start, next)) > max_tokens
        {
            output.push(message_ref(message, piece_start, cursor));
            piece_start = cursor;
        }
        cursor = next;
    }
    if piece_start < end {
        output.push(message_ref(message, piece_start, end));
    }
    output
}

fn message_ref(message: &NormalizedMessage, start: usize, end: usize) -> MessageRef {
    MessageRef {
        message_id: message.id.clone(),
        start_char: start,
        end_char: end,
        text: char_slice(&message.text, start, end),
    }
}

fn char_slice(text: &str, start: usize, end: usize) -> String {
    text.chars().skip(start).take(end - start).collect()
}

fn pack_ranges(refs: &[MessageRef], max_tokens: usize) -> Vec<(usize, usize)> {
    let mut groups = Vec::new();
    let mut start = 0;
    let mut tokens = 0;
    for (index, reference) in refs.iter().enumerate() {
        let next = estimate_tokens(&reference.text);
        if index > start && tokens + next > max_tokens {
            groups.push((start, index));
            start = index;
            tokens = 0;
        }
        tokens += next;
    }
    if start < refs.len() {
        groups.push((start, refs.len()));
    }
    groups
}

fn select_before(refs: &[MessageRef], limit: usize) -> Vec<MessageRef> {
    if limit == 0 {
        return Vec::new();
    }
    let mut selected = Vec::new();
    let mut seen = Vec::<String>::new();
    for reference in refs.iter().rev() {
        if !seen.contains(&reference.message_id) {
            if seen.len() >= limit {
                break;
            }
            seen.push(reference.message_id.clone());
        }
        selected.push(reference.clone());
    }
    selected.reverse();
    selected
}

fn select_after(refs: &[MessageRef], limit: usize) -> Vec<MessageRef> {
    if limit == 0 {
        return Vec::new();
    }
    let mut selected = Vec::new();
    let mut seen = Vec::<String>::new();
    for reference in refs {
        if !seen.contains(&reference.message_id) {
            if seen.len() >= limit {
                break;
            }
            seen.push(reference.message_id.clone());
        }
        selected.push(reference.clone());
    }
    selected
}

fn trim_context(
    mut before: Vec<MessageRef>,
    candidate: &[MessageRef],
    mut after: Vec<MessageRef>,
    max_tokens: usize,
) -> (Vec<MessageRef>, Vec<MessageRef>) {
    while refs_tokens(before.iter().chain(candidate).chain(after.iter())) > max_tokens {
        if !before.is_empty() {
            before.remove(0);
        } else if !after.is_empty() {
            after.pop();
        } else {
            break;
        }
    }
    (before, after)
}

fn refs_tokens<'a>(refs: impl Iterator<Item = &'a MessageRef>) -> usize {
    refs.map(|reference| estimate_tokens(&reference.text)).sum()
}

fn make_window(
    episode: &ConversationEpisode,
    candidate: Vec<MessageRef>,
    before: Vec<MessageRef>,
    after: Vec<MessageRef>,
    config: &WindowConfig,
) -> ExtractionWindow {
    let id = format!(
        "window-{}",
        stable_hash(&[
            json!(episode.scope_id),
            json!(episode.session_id),
            to_value(&candidate).expect("serializable refs"),
            to_value(&before).expect("serializable refs"),
            to_value(&after).expect("serializable refs"),
            to_value(config).expect("serializable config"),
        ])
    );
    let mut seen = HashSet::new();
    let candidate_message_ids = candidate
        .iter()
        .filter(|reference| seen.insert(reference.message_id.clone()))
        .map(|reference| reference.message_id.clone())
        .collect();
    let candidate_token_count = refs_tokens(candidate.iter());
    let total_token_count = refs_tokens(before.iter().chain(candidate.iter()).chain(after.iter()));
    ExtractionWindow {
        id,
        scope_id: episode.scope_id.clone(),
        session_id: episode.session_id.clone(),
        episode_id: episode.id.clone(),
        candidate_refs: candidate,
        context_before_refs: before,
        context_after_refs: after,
        candidate_message_ids,
        candidate_token_count,
        total_token_count,
        window_version: config.version.clone(),
    }
}
